staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filter.go

```diff
@@ -0,0 +1,273 @@
+	serverConcurrencyLimit int,
+	requestWaitLimit time.Duration,
+	waitForAllPredefined bool,
+) Interface {
+	wg := counter.NoOp{}
+	return NewRequestManagerTestable(
+		informerFactory,
+		flowcontrolClient,
+		serverConcurrencyLimit,
+		requestWaitLimit,
+		waitForAllPredefined,
+		wg,
+		fqs.NewQueueSetFactory(&clock.RealClock{}, wg),
+	)
+}
+
+// NewRequestManagerTestable is extra flexible to facilitate testing
+func NewRequestManagerTestable(
+	informerFactory kubeinformers.SharedInformerFactory,
+	flowcontrolClient rmclientv1alpha1.FlowcontrolV1alpha1Interface,
+	serverConcurrencyLimit int,
+	requestWaitLimit time.Duration,
+	waitForAllPredefined bool,
+	wg counter.GoRoutineCounter,
+	queueSetFactory fq.QueueSetFactory,
+) Interface {
+	reqMgr := &requestManager{
+		wg:                     wg,
+		queueSetFactory:        queueSetFactory,
+		serverConcurrencyLimit: serverConcurrencyLimit,
+		requestWaitLimit:       requestWaitLimit,
+		flowcontrolClient:      flowcontrolClient,
+	}
+	klog.V(2).Infof("NewRequestManagementSystem with serverConcurrencyLimit=%d, requestWaitLimit=%s", serverConcurrencyLimit, requestWaitLimit)
+	reqMgr.initializeConfigController(informerFactory)
+	emptyRMState := &requestManagerState{
+		priorityLevelStates: make(map[string]*priorityLevelState),
+	}
+	reqMgr.curState.Store(emptyRMState)
+	reqMgr.digestConfigObjects(nil, nil)
+	return reqMgr
+}
+
+// RequestDigest holds necessary info from request for flow-control
+type RequestDigest struct {
+	RequestInfo *request.RequestInfo
+	User        user.Info
+}
+
+func (reqMgr *requestManager) Wait(ctx context.Context, requestDigest RequestDigest) (bool, func()) {
+	startWaitingTime := time.Now()
+	for {
+		rmState := reqMgr.curState.Load().(*requestManagerState)
+
+		// 1. figure out which flow schema applies
+		fs := rmState.pickFlowSchema(requestDigest)
+		if fs == nil { // reject
+			metrics.AddReject("<none>", "non-match")
+			klog.V(7).Infof("Rejecting requestInfo=%#+v, userInfo=%#+v because no FlowSchema matched", requestDigest.RequestInfo, requestDigest.User)
+			return false, func() {}
+		}
+		plName := fs.Spec.PriorityLevelConfiguration.Name
+
+		// 2. early out for exempt
+		ps := rmState.priorityLevelStates[plName]
+		if ps.config.Type == rmtypesv1alpha1.PriorityLevelEnablementExempt {
+			klog.V(7).Infof("Serving requestInfo=%#+v, userInfo=%#+v, fs=%s, pl=%s without delay", requestDigest.RequestInfo, requestDigest.User, fs.Name, plName)
+			startExecutionTime := time.Now()
+			return true, func() {
+				metrics.ObserveExecutionDuration(plName, fs.Name, time.Now().Sub(startExecutionTime))
+			}
+		}
+
+		// 3. computing hash
+		flowDistinguisher := requestDigest.ComputeFlowDistinguisher(fs.Spec.DistinguisherMethod)
+		hashValue := hashFlowID(fs.Name, flowDistinguisher)
+
+		// 4. queuing
+		tryAnother, execute, afterExecute := ps.queues.Wait(ctx, hashValue, requestDigest.RequestInfo, requestDigest.User)
+		if tryAnother {
```

Comment:
I also tend to think that a read/write lock would have been a fine choice.
* Split the state into two copies, the state used by this enqueueing code, and the state used by the dispatching code.
* In this function, getting the state and selecting a qs, would be done while holding the read lock.
* In the dispatching code, a (possibly separate) read lock is held while selecting a request to consume.
* The config changing code takes a write lock to make modifications. It leaves non-empty queuesets in the dispatching state until they become empty, but removes them from the enqueueing state as soon as the config stops referencing them.

In this possible world, we don't need the tryAnother concept. We do mutations under a write lock instead of making copies. The write lock only blocks a small piece of the request handling pathway, and we can do pre-work (e.g., construct the new queuesets, assemble the list that needs to be deleted) to minimize the amount of time the write lock is held. I think the code would be quite a bit simpler and easier to reason about. Basically, if it were written that way, and someone proposed a change to the current form, I think we'd demand a profile demonstrating e.g. a 5x improvement in lock wait times or something like that.

We've already litigated this and although I don't think it should stay in its current form, I'm not asking for this to change in this PR.

Reply:
> so that's another thing that can get simpler if we add a lock. 

IIUC we will have to add a new list of the dangling queues from deleted PLs to apply r/wlock here. hmm, doesnt sounds like a simpler one comparing to the "try-another" approach.

Reply:
Clever and simple are opposites. The current locking method here is extremely clever. The thing @jktomer and I are suggesting is very stupid, but it would be much more obvious that it works when one reads the code. Code that "obviously" doesn't have any bugs and is also tested is the gold standard. It is easier to get the obviousness property with "stupid looking" code than with clever code.

In particular, having fewer thingies in the code doesn't necessarily make code simpler. In fact, if you are doing the same work with fewer named entities (that is, shorter code), then the code necessarily must be more [complex](https://en.wikipedia.org/wiki/Kolmogorov_complexity)!

Sometimes shorter code is easier for humans to understand, and sometimes it is harder to understand. Locality (how many functions/objects do you have to keep in mind, and how far away are they) is more important than length.

The code we write is a living artifact, and others will change it, and *many* others will *read* it to understand the system behavior. We need to make their lives as easy as possible. We want to spend our cleverness points on the overall design, not on the code's implementation.

I can't really argue with "hmm, doesnt sounds like a simpler one" since the way things sound to you is not really up for debate. So, instead, I hope this helps you understand more which property I want the code to have. Whether you understand this concept as "simplicity" or want to call it something else, I don't mind too much.